### PR TITLE
Bump pytorch to 2.5.0

### PR DESCRIPTION
**Pull request by b-chu**
# What does this PR do?

Bumps pytorch version to 2.5.0

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: .github/workflows/pr-cpu.yaml [hunk 1]
```diff
@@ -16,10 +16,6 @@ jobs:
     strategy:
       matrix:
         include:
-        - name: cpu-3.11-2.2
```
Comment (mvpatel2000) [inline .github/workflows/pr-cpu.yaml:19]: you will need to similarly remove GPU jobs for 2.2

File: composer/checkpoint/state_dict.py [hunk 2]
```diff
@@ -88,7 +88,7 @@ def get_model_state_dict(
             log.debug('Calling model.state_dict() for non-FSDP model...')
             model_state_dict = model.state_dict()
         if isinstance(model, DistributedDataParallel):
-            nn.modules.utils.consume_prefix_in_state_dict_if_present(model_state_dict, 'module.')
+            nn.modules.utils.consume_prefix_in_state_dict_if_present(model_state_dict, 'module.')  # type: ignore
```
Comment (mvpatel2000) [inline composer/checkpoint/state_dict.py:91]: why is this needed? are you sure this isnt API change?

File: composer/models/huggingface.py [hunk 3]
```diff
@@ -916,7 +916,7 @@ def write_huggingface_pretrained_from_composer_checkpoint(
         peft_config.save_pretrained(str(output_folder))
 
     weights_state_dict = composer_state_dict['state']['model']
-    torch.nn.modules.utils.consume_prefix_in_state_dict_if_present(weights_state_dict, prefix='model.')
+    torch.nn.modules.utils.consume_prefix_in_state_dict_if_present(weights_state_dict, prefix='model.')  # type: ignore
```
Comment (mvpatel2000) [inline composer/models/huggingface.py:919]: same comment

File: docker/build_matrix.yaml [hunk 4]
```diff
@@ -2,129 +2,103 @@
...
 - AWS_OFI_NCCL_VERSION: ''
-  BASE_IMAGE: nvidia/cuda:12.1.1-cudnn8-devel-ubuntu20.04
-  CUDA_VERSION: 12.1.1
-  IMAGE_NAME: torch-2-3-1-cu121
+  BASE_IMAGE: nvidia/cuda:12.4.1-cudnn-devel-ubuntu20.04
+  CUDA_VERSION: 12.4.1
+  IMAGE_NAME: torch-2-4-1-cu124
   MOFED_VERSION: latest-23.10
-  NVIDIA_REQUIRE_CUDA_OVERRIDE: cuda>=12.1 brand=tesla,driver>=450,driver<451 brand=tesla,driver>=470,driver<471
-    brand=unknown,driver>=470,driver<471 brand=nvidia,driver>=470,driver<471 brand=nvidiartx,driver>=470,driver<471
```
Comment (mvpatel2000) [inline docker/build_matrix.yaml:59]: iirc these are needed and cannot be removed

File: tests/checkpoint/helpers.py [hunk 5]
```diff
@@ -8,7 +8,7 @@
 from packaging import version
 from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
 from torch.distributed.fsdp.api import CPUOffload
-from torch.optim import adam
+from torch.optim import Adam
```
Comment (mvpatel2000) [inline tests/checkpoint/helpers.py:11]: is this backwards compatible? eg does this import work on torch 2.3?

**Review by mvpatel2000 — COMMENTED** [review]
you will need to update `patch_pytorch` as well, e.g for the sync hook for auto microbatching

File: docker/build_matrix.yaml (hunk 4 above)
Comment (b-chu) [reply docker/build_matrix.yaml:59]: this was auto generated though?

File: tests/checkpoint/helpers.py (hunk 5 above)
Comment (b-chu) [reply tests/checkpoint/helpers.py:11]: Yes

Comment (b-chu) [reply tests/checkpoint/helpers.py:11]: (otherwise it wouldn't have passed 2.3 tests also)

File: .github/workflows/pr-cpu.yaml (hunk 1 above)
Comment (b-chu) [reply .github/workflows/pr-cpu.yaml:19]: Yes, in progress

File: composer/checkpoint/state_dict.py (hunk 2 above)
Comment (b-chu) [reply composer/checkpoint/state_dict.py:91]: It is an API change (they've been adding `__all__` to their files) but they didn't expose their utils correctly

File: docker/build_matrix.yaml (hunk 4 above)
Comment (mvpatel2000) [reply docker/build_matrix.yaml:59]: Yes

File: docker/generate_build_matrix.py [hunk 6]
```diff
@@ -180,7 +180,7 @@ def _write_table(table_tag: str, table_contents: str):
 
 
 def _main():
-    python_pytorch_versions = [('3.11', '2.4.1'), ('3.11', '2.3.1'), ('3.11', '2.2.2')]
+    python_pytorch_versions = [('3.11', '2.5.0'), ('3.11', '2.4.1'), ('3.11', '2.3.1')]
```
Comment (mvpatel2000) [inline docker/generate_build_matrix.py:183]:
```suggestion
    python_pytorch_versions = [('3.12', '2.5.0'), ('3.11', '2.4.1'), ('3.11', '2.3.1')]
```
3.12?

File: composer/trainer/_patch_pytorch.py [hunk 7]
```diff
@@ -109,6 +109,12 @@ def patch_pytorch():
         # No monkeypatches!
         pass
 
+    elif version.parse(torch.__version__) < version.parse('2.5.1'):
+        # Monkey patch for torch < 2.5.1 ie torch == 2.5.0
+
+        # No monkeypatches!
```
Comment (mvpatel2000) [inline composer/trainer/_patch_pytorch.py:115]: we should prob make htis comment more clear that says besides the ones for unshard

**Review by mvpatel2000 — COMMENTED** [review]
what about the SGD and ckpt changes?

File: composer/trainer/trainer.py [hunk 8]
```diff
@@ -2317,9 +2317,12 @@ def fit(
...
 
         # Raise error if callig fit with SGD
-        if type(
-            self.state.optimizers[0],
-        ) == torch.optim.SGD and version.parse(torch.__version__) >= version.parse('2.4.0'):
+        if (
+            type(
+                self.state.optimizers[0],
+            ) == torch.optim.SGD and version.parse(torch.__version__) >= version.parse('2.4.0') and
+            version.parse(torch.__version__) >= version.parse('2.4.0')
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:2324]: should be <2.5

File: composer/trainer/trainer.py [hunk 9]
```diff
@@ -2317,9 +2317,12 @@ def fit(
             self.state.max_duration = duration + self.state.timestamp.get(duration.unit)
 
         # Raise error if callig fit with SGD
-        if type(
-            self.state.optimizers[0],
-        ) == torch.optim.SGD and version.parse(torch.__version__) >= version.parse('2.4.0'):
+        if (
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:2320]: this linting is so sad

File: composer/trainer/trainer.py (hunk 8 above)
Comment (b-chu) [reply composer/trainer/trainer.py:2324]: Weird, coulda sworn I changed this

File: composer/trainer/trainer.py (hunk 9 above)
Comment (b-chu) [reply composer/trainer/trainer.py:2320]: Fine I fixed it

File: composer/trainer/_patch_pytorch.py (hunk 7 above)
Comment (b-chu) [reply composer/trainer/_patch_pytorch.py:115]: Done

File: docker/generate_build_matrix.py (hunk 6 above)
Comment (b-chu) [reply docker/generate_build_matrix.py:183]: Approve 2.5.0 first then I'll make 3.12 changes (same PR) and see how many things it breaks

**Comment by bigning** [comment]
can you run a daily test ?

**Review by mvpatel2000 — COMMENTED** [review]
LGTM. Let's do python bump

**Comment by b-chu** [comment]
All daily tests pass
https://github.com/mosaicml/composer/actions/runs/11412854427

**Comment by b-chu** [comment]
Moving PR to #3663
